src/java/org/apache/cassandra/io/sstable/format/SSTableReaderBuilder.java

```diff
@@ -381,11 +401,10 @@ private void load(ValidationMetadata validation,
                           DiskOptimizationStrategy optimizationStrategy,
                           StatsMetadata statsMetadata) throws IOException
         {
-            if (metadata.params.bloomFilterFpChance == 1.0)
+            if (!BloomFilter.shouldUseBloomFilter(metadata.params.bloomFilterFpChance))
             {
                 // bf is disabled.
                 load(false, !isOffline, optimizationStrategy, statsMetadata, components);
-                bf = FilterFactory.AlwaysPresent;
```

Comment:
Yes, it seems that your are right about this branch - in this case, we explicitly want to set it to pass-through filter. 

There is a possibility that the filter is set manually using a setter method `bf` on the builder. I don't know what should be the logic then though

Reply:
We could probably prevent any loading / recreating of BF if it is externally set - but tbh I can see it is only set from the ForWriter - so maybe that particular methods should be moved there

Reply:
So I've moved those methods which explicitly modify fields to `ForWriter` impl because they are used only in that context. `ForWriter` never loads `bf` and it relies only on `bf` set externally. Therefore, whenever we use `load` method, `bf` can only be `null` when we enter the method and we should never allow for passing `null` `bf` to `SSTableReader` as nothing is expecting it would ever be `null`

Reply:
Ah, ok, as I thought it was mostly that I didn't quite understand what was happening, but I think this makes it more clear.